src/main/java/org/medicmobile/webapp/mobile/SettingsStore.java

```diff
@@ -45,13 +45,34 @@ void updateWithUnlockCode(String unlockCode) throws SettingsException {
 		ed.putString("unlock-code", unlockCode);
 
 		if(!ed.commit()) throw new SettingsException(
-				"Failed to save to SharedPreferences.");
+				"Failed to save 'unlock-code' to SharedPreferences.");
 	}
 
 	String get(String key) {
 		return prefs.getString(key, null);
 	}
 
+	/**
+	 * Returns true if the user has denied to provide its geolocation data.
+	 * The rejection is taken from the first dialog with the "prominent" disclosure
+	 * about the location data, not from the native dialog displayed by Android.
+	 *
+	 * @see RequestPermissionDialog
+	 */
+	boolean hasUserDeniedGeolocation() {
+		return prefs.getBoolean("denied-geolocation", false);
+	}
+
+	/**
+	 * @see #hasUserDeniedGeolocation()
+	 */
+	void setUserDeniedGeolocation() throws SettingsException {
+		SharedPreferences.Editor ed = prefs.edit();
+		ed.putBoolean("denied-geolocation", true);
+		if(!ed.commit()) throw new SettingsException(
+				"Failed to save 'denied-geolocation' to SharedPreferences.");
+	}
+
```

Comment:
I'm not sure we should store this in the settings. I think this would be very difficult to clear if the user has changed their mind - we'd either have to add a UI to do it or wipe all app data.

Android already remembers this information and provides a UI for changing the permissions. Can we instead check that somehow and only show the dialog if it's not already set?

Reply:
Yes it is something discussed in the ticket, in the current master version if you regret to allow / deny the locations, you have the same problem, but it is true that it is more convenient just to remove the permission or the rejection from the app info in the Android settings than wipe all the data in order to clean this setting, but either case the user needs to go to the Android settings to do so.

> Can we instead check that somehow and only show the dialog if it's not already set?

It works like that, first it checks whether the user already allowed / denied the locations permissions, and only shows the dialog if the user never did. Users that already has the app installed and already allowed the app to access the location won't see the new dialog when upgrade to this version, and the app won't store any configuration.

Anyway, I'll check later if there is a way that in case the user rejects the permission request from our dialog, store the rejection in the same permissions settings that Android uses for that purpose, but because it is a rejection, not sure whether is possible.

Reply:
Back to this, I pushed a fix that was causing the form to be blocked when loc permissions were denied, and now testing again, turns out the user does not need to wipe the app data to rollback its decision to allow or deny the access to the location, it only needs to do the same that it does with the current master version: go to the app settings in the android settings and just add again or revoke the location permission, but no need to wipe the app data because the variable stored here is read after check the permissions against the Android API.

The only difference is that in case the user rejected the location permissions from the custom dialog instead of from the Android dialog, in the app settings you won't see the location permission accepted or rejected, the permission will be displayed like it was never accepted or rejected, and from there the user will be able to add the permission to access location in which case this variable stored will be ignored by the app.

Still will check later whether I can ask Android to store the negative of location access using the API like it was rejected from the native location dialog, but may be not possible.